File: services/php/src/tests/json-serde.php
<?php

/**
 * Prints MD5 hash of data
 */
function printHash($data, $label = ''): string
{
    $str = json_encode($data);
    $hash = md5($str);
    return ($label ? $label . ' ' : '') . 'MD5 Hash: ' . $hash;
}

/**
 * Performs JSON serialization/deserialization benchmark
 */
function json_serde(): string
{
    $sampleData = [
        'users' => [
            ['id' => 1, 'name' => 'Alice', 'email' => 'alice@example.com', 'active' => true],
            ['id' => 2, 'name' => 'Bob', 'email' => 'bob@example.com', 'active' => false],
            ['id' => 3, 'name' => 'Charlie', 'email' => 'charlie@example.com', 'active' => true]
        ],
        'metadata' => [
            'version' => '1.0.0',
            'timestamp' => date('c'),
            'settings' => [
                'theme' => 'dark',
                'notifications' => true,
                'language' => 'en'
            ]
        ],
        'statistics' => [
            'totalUsers' => 3,
            'activeUsers' => 2,
            'averageAge' => 28.5,
            'tags' => ['javascript', 'nodejs', 'benchmark', 'json', 'serialization']
        ]
    ];

    $n = 100; // Number of serialization/deserialization cycles

    $jsonStr = json_encode($sampleData);
    $parsed = json_decode($jsonStr, true);
    $originalHash = printHash($parsed, 'Original');

    $results = [];
    for ($i = 0; $i < $n; $i++) {
        $serialized = json_encode($parsed);
        $deserialized = json_decode($serialized, true);
        $results[] = $deserialized;
    }

    $finalHash = printHash($results, "x{$n} Cycles");

    return implode("\n", [
        $originalHash,
        $finalHash,
        "Performed {$n} serialization/deserialization cycles",
        'Original data size: ' . strlen($jsonStr) . ' bytes',
        'Final array size: ' . count($results) . ' objects'
    ]);
}

?>

File: services/python/src/tests/json_serde.py
import json
import hashlib
from datetime import datetime

def print_hash(data, label=''):
    """Calculate and format MD5 hash of JSON-serialized data"""
    json_str = json.dumps(data, sort_keys=True)
    hash_obj = hashlib.md5(json_str.encode('utf-8'))
    hash_hex = hash_obj.hexdigest()
    prefix = f"{label} " if label else ""
    return f"{prefix}MD5 Hash: {hash_hex}"

def json_serde():
    """Perform JSON serialization/deserialization cycles"""
    sample_data = {
        "users": [
            {"id": 1, "name": "Alice", "email": "alice@example.com", "active": True},
            {"id": 2, "name": "Bob", "email": "bob@example.com", "active": False},
            {"id": 3, "name": "Charlie", "email": "charlie@example.com", "active": True}
        ],
        "metadata": {
            "version": "1.0.0",
            "timestamp": datetime.now().isoformat(),
            "settings": {
                "theme": "dark",
                "notifications": True,
                "language": "en"
            }
        },
        "statistics": {
            "totalUsers": 3,
            "activeUsers": 2,
            "averageAge": 28.5,
            "tags": ["javascript", "nodejs", "benchmark", "json", "serialization"]
        }
    }

    n = 100  # Number of serialization/deserialization cycles

    json_str = json.dumps(sample_data)
    parsed = json.loads(json_str)
    original_hash = print_hash(parsed, 'Original')

    results = []
    for i in range(n):
        serialized = json.dumps(parsed)
        deserialized = json.loads(serialized)
        results.append(deserialized)

    final_hash = print_hash(results, f'x{n} Cycles')

    return '\n'.join([
        original_hash,
        final_hash,
        f'Performed {n} serialization/deserialization cycles',
        f'Original data size: {len(json_str)} bytes',
        f'Final array size: {len(results)} objects'
    ])

